### ISSUE TITLE
Формы в модуле User

### ISSUE THREAD

**Issue by EvanLoik** [yupe/yupe#1915]
Было бы неплохо сделать так, чтобы формы (например RegistrationForm) наследовались от модели User. Сейчас они наследуются от CFormModel и из-за этого приходится во всех формах прописывать AttributeLabels и необходимые переменные. Это крайне неудобно, если возникает необходимость модифицировать модуль User.

**Comment by yupe** [comment]
Форма не может быть унаследована от модели. Это две разные сущности.

**Comment by EvanLoik** [comment]
Вы думаете? В Yii2, например, стандартная реализация формы поиска(через gii) осуществлена с помощью наследования от модели. Т.е есть модель Post, а есть PostSearch, наследуемая от модели Post. В ней переопределяются необходимые методы и добавляются свои правила валидации. Вы не считаете, что так удобнее? А то сейчас, для того, чтобы добавить новое поле необходимо изменить и RegistrationForm, и ProfileForm, и модель User. И если, не дай бог, вы случайно ошиблись с названием поля - все полетит к чертям =))

**Comment by yupe** [comment]
Мне не нравится подход в уии2. Форма - это форма. Все свойства формы прописаны явно как атрибуты класса формы. В моделе атрибуты подтягиваются из схемы базы данных. Не вижу смысла смешивать чисто формы и модели.

**Comment by chemezov** [comment]
@yupe почему не может? Можно унаследоваться от User и полностью переопределить rules(). Получится такая же форма. Другое дело, что сверху ещё подтянется дофига всего.

Можно поступить так: создать BaseUser прямо как если бы модель была только что сгенерирована через gii. А User уже наследовать от неё, со всей кастомной логикой. Ну и RegistrationForm можно попробовать расширить от BaseUser. Самому надоедает добавлять одни и те же поля в несколько форм.

В Yii2 кстати удачно сделано, через common/backend/frontend. Пока проект/модуль небольшой, не замечаешь выгоды, зато потом...

**Comment by DarkCs** [comment]
Post и PostSearch это неудачный пример, RegistrationForm и User это две разные сущности, то что у них много общих полей - просто совпадение

**Comment by yupe** [comment]
@chemezov  не стоит путать формы и модели. Модель - это сущность предметной области. Форма - набор полей и не более. Одна форма может содержать данные для больше чем одной модели. Не вижу ничего страшного в добавлении атрибута в несколько форм.

**Comment by EvanLoik** [comment]
Да, я вас понял. Спасибо за замечания. Тогда у меня другой вопрос: зачем указывать правила валидации в методе rules() в модели User? Ведь у каждой формы свои правила. Вы не находите, что это неудобно, прописывать правила как в модели, так и в форме? Можно в методе rules() вернуть пустой массив (в модели), а в каждой форме прописать свои правила. Сейчас у Вас есть правила, которые повторяются и формах, и в модели. Может стоит подумать над тем, как избежать дублирования кода? Все-таки на практике это неудобно.

**Comment by yupe** [comment]
Во-первых в Yii нет красивого способа отделить валидацию от модели. Можно посмотреть как это сделано в Laravel или Symfony. Во-вторых в формах есть далеко не все поля модели User. В-третьих модели используются минуя формы в панели управления, кроме того модели без форм могут использоваться, например, в консольных приложениях. Валидация бизнес-логики должна быть в независимости от того использовалась форма или нет.

**Comment by EvanLoik** [comment]
@yupe Хорошо. Тогда зачем нужна повторная валидация в модели(в конкретных случаях), если уже есть валидация в форме?
Например, в методе createUser компонента UserManager вы сделали так:
```php
 $user->setAttributes($data);
```
а потом так:
```php
 if ($user->save() && ($token = $this->tokenStorage->createAccountActivationToken($user)) !== false)
```

Может имеет смысл сделать так:
```php
$user->setAttributes($data, false);
```
и
```php
$user->save(false)
```
Таким образом, валидацию будет осуществлять конкретно RegistrationForm, без модели User, при этом сохраняя и правила самой модели, на тот случай, если не нужно использовать форму.

**Comment by yupe** [comment]
@EvanLoik валидация при сохранении данных должна быть в любом случае. Форма регистрации может отличаться от стандартной (вынесли "пол" в регистрацию и сделали свою форму, например). Валидация у форм предназначена чтобы всего лишь показать (или нет) сообщения об ошибках на уровне представления. Это не отменяет необходимости валидации данных при сохранении данных.

**Comment by EvanLoik** [comment]
@yupe Тогда зачем делать прописывать валидатор filter в формах? Зачем это делать два раза? Получается, чтобы создать новое поле в форме нужно:
1. Прописать атрибут класса RegisterForm
2. Прописать атрибут класса ProfileForm
3. Прописать его в RegisterForm::rules()
4. Прописать его в RegisterForm::attributeLabels()
5. Прописать его в User::rules()
6. Прописать его в User::attributeLabels()
7. Прописать его в ProfileForm::rules()
8. Прописать его в ProfileForm::attributeLabels()
9. Убедиться, что эти правила никак не конфликтуют между собой. Потому, что если они где-то отличаются - в некоторых случаях система просто выбьет нам Error creating account!

Прошу заметить, что есть много правил, таких, как filter, которые, на мой взгляд, не нужно использовать два раза. Ведь в этом нету смысла.

Сейчас, например, для того, чтобы просто сделать атрибут nick_name необязательным необходимо подпортить себе нервную систему. Тем более, если его надо удалить (а авторизацию сделать по e-mail).

Как вы знаете, в каждом проекте по-любому нужно или удалять какие-то поля, или добавлять. И такая реализация, на мой взгляд, не практичная. В том же Yii2, в примере, который скинул @DarkCs все логично. Там форма валидирует присущие ей поля. А в модели указаны общие правила, для всех форм. 

Я полностью согласен с тем, что форма и модель - это две разные сущности. Прошу прощения, я сразу об этом не подумал. Но все равно, надо подумать над тем, как упростить процесс модификации модуля User. Вас самого не смущает эта проблема? И, если ничего из предложенного выше Вас не устраивает, может у Вас есть какие-то мысли? Ведь согласитесь, разработчику должно быть удобно работать с Юпи. И нет ничего хорошего в том, чтобы дублировать код и усложнять работу без надобности.

**Comment by maxim-avramenko** [comment]
Есть модель User и есть куча форм для работы с моделью, например RegisterForm, UpdateUserForm и т.д 
Модель User имеет свои правила валидации, а формы имеют другие правила валидации, они могут совпадать, а могут и отличаться. Если данные из RegisterForm прошли валидацию то повторная валидация в модельке User не имеет смысла, поэтому смело сохраняем без валидации $user->save(false); и едем дальше. НО! тут необходимо помнить о том что бы форма RegistrationForm содержала валидацию полей полностью покрывающую все требования модели User, иначе дальше моделька User не сможет сохраниться если где то дернут $user->save(); без false и окажется что часть данных в модельке не валидна. Поэтому формы для User должны иметь много общего с моделькой, включая имена свойств (если нужно массовое заполнение/присваивание полей) и правила валидации (что бы модель User была все таки валидной перед сохранением). Это не только в Yii и Yii2, это можно встретить в любом другом  framework :)
